core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -501,9 +521,8 @@ public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
     return new Builder(this).setCurrentSnapshot(snapshot).build();
   }
 
-  public TableMetadata removeSnapshotsIf(Predicate<Snapshot> removeIf) {
-    List<Snapshot> toRemove = snapshots.stream().filter(removeIf).collect(Collectors.toList());
-    return new Builder(this).removeSnapshots(toRemove).build();
+  public TableMetadata removeSnapshots(Set<Long> snapshotsToRemove) {
```

Comment:
I also created https://github.com/apache/iceberg/issues/3900 to track the new API. Not sure what is the preference, but I think it's better to have a PR that is dedicated to the new APIs and old method deprecation, because this PR is already quite long.